### PR TITLE
Team Connect

### PR DESCRIPTION
**Pull request by AustinShalit**
2 Problems I need to solve:
- [ ] Dialog starts small (#124)
- [x] when changing the port number, NetworkTables does not start again (I am calling initialize but it does not start).  Update: This is a bug in ntcore.

### CODE REVIEW

File: app/src/main/java/edu/wpi/first/shuffleboard/app/MainWindowController.java [hunk 1]
```diff
@@ -289,27 +286,23 @@ public void showPrefs() {
...
     AppPreferences.getInstance().getProperties()
         .stream()
-        .map(property -> new ObservableItem(property, "Application"))
+        .map(property -> new FlushableItem(property, "Application"))
         .forEachOrdered(propertySheet.getItems()::add);
     propertySheet.setPropertyEditorFactory(new PropertyEditorFactory());
-    StackPane pane = new StackPane(propertySheet);
-    pane.setPadding(new Insets(8));
-    Scene scene = new Scene(pane);
-    EasyBind.listBind(scene.getRoot().getStylesheets(), root.getStylesheets());
```
Comment (SamCarlberg) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/MainWindowController.java]: This line is needed to keep the application style on the prefs dialog

Comment (AustinShalit) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/MainWindowController.java]: Added back.

**Comment by SamCarlberg** [comment]
When opening the preferences pane:

```java
Exception in thread "JavaFX Application Thread" java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at javafx.fxml.FXMLLoader$MethodHandler.invoke(FXMLLoader.java:1774)
	at javafx.fxml.FXMLLoader$ControllerMethodEventHandler.handle(FXMLLoader.java:1657)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.control.MenuItem.fire(MenuItem.java:462)
	at com.sun.javafx.scene.control.ControlAcceleratorSupport.lambda$doAcceleratorInstall$12(ControlAcceleratorSupport.java:165)
	at com.sun.javafx.scene.KeyboardShortcutsHandler.processAccelerators(KeyboardShortcutsHandler.java:347)
	at com.sun.javafx.scene.KeyboardShortcutsHandler.dispatchBubblingEvent(KeyboardShortcutsHandler.java:163)
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.Scene$KeyHandler.process(Scene.java:3964)
	at javafx.scene.Scene$KeyHandler.access$1800(Scene.java:3910)
	at javafx.scene.Scene.impl_processKeyEvent(Scene.java:2040)
	at javafx.scene.Scene$ScenePeerListener.keyEvent(Scene.java:2501)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(GlassViewEventHandler.java:217)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(GlassViewEventHandler.java:149)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleKeyEvent$353(GlassViewEventHandler.java:248)
	at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:389)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleKeyEvent(GlassViewEventHandler.java:247)
	at com.sun.glass.ui.View.handleKeyEvent(View.java:546)
	at com.sun.glass.ui.View.notifyKey(View.java:966)
	at com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method)
	at com.sun.glass.ui.gtk.GtkApplication.lambda$null$49(GtkApplication.java:139)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.reflect.misc.Trampoline.invoke(MethodUtil.java:71)
	at sun.reflect.GeneratedMethodAccessor1.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.reflect.misc.MethodUtil.invoke(MethodUtil.java:275)
	at javafx.fxml.FXMLLoader$MethodHandler.invoke(FXMLLoader.java:1771)
	... 36 more
Caused by: java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.Double
	at edu.wpi.first.shuffleboard.app.widget.WidgetPropertySheet$NumberPropertyEditor.setValue(WidgetPropertySheet.java:163)
	at impl.org.controlsfx.skin.PropertySheetSkin$PropertyPane.getEditor(PropertySheetSkin.java:346)
	at impl.org.controlsfx.skin.PropertySheetSkin$PropertyPane.setItems(PropertySheetSkin.java:301)
	at impl.org.controlsfx.skin.PropertySheetSkin$PropertyPane.<init>(PropertySheetSkin.java:269)
	at impl.org.controlsfx.skin.PropertySheetSkin$PropertyPane.<init>(PropertySheetSkin.java:261)
	at impl.org.controlsfx.skin.PropertySheetSkin.buildPropertySheetContainer(PropertySheetSkin.java:223)
	at impl.org.controlsfx.skin.PropertySheetSkin.refreshProperties(PropertySheetSkin.java:188)
	at impl.org.controlsfx.skin.PropertySheetSkin.<init>(PropertySheetSkin.java:143)
	at org.controlsfx.control.PropertySheet.createDefaultSkin(PropertySheet.java:254)
	at javafx.scene.control.Control.impl_processCSS(Control.java:872)
	at javafx.scene.Parent.impl_processCSS(Parent.java:1280)
	at javafx.scene.Node.processCSS(Node.java:9058)
	at javafx.scene.Scene.doCSSPass(Scene.java:545)
	at javafx.scene.Scene.preferredSize(Scene.java:1643)
	at javafx.scene.Scene.impl_preferredSize(Scene.java:1720)
	at javafx.stage.Window$9.invalidated(Window.java:846)
	at javafx.beans.property.BooleanPropertyBase.markInvalid(BooleanPropertyBase.java:109)
	at javafx.beans.property.BooleanPropertyBase.set(BooleanPropertyBase.java:144)
	at javafx.stage.Window.setShowing(Window.java:922)
	at javafx.stage.Window.show(Window.java:937)
	at javafx.stage.Stage.show(Stage.java:259)
	at javafx.stage.Stage.showAndWait(Stage.java:472)
	at javafx.scene.control.HeavyweightDialog.showAndWait(HeavyweightDialog.java:162)
	at javafx.scene.control.Dialog.showAndWait(Dialog.java:341)
	at edu.wpi.first.shuffleboard.app.MainWindowController.showPrefs(MainWindowController.java:297)
	... 46 more
```

**Comment by AustinShalit** [comment]
Fixed now.  This is why tests would be good.

File: api/src/main/java/edu/wpi/first/shuffleboard/api/components/IntegerField.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+  /**
+   * Creates a new number field with no value.
+   */
+  public IntegerField() {
+    super();
+    setText("0"); // initial text to match the initial number
+    setTextFormatter(new TextFormatter<>(change -> {
+      String text = change.getControlNewText();
+      if (isNumber(text)) {
```
Comment (nightpool) [inline api/src/main/java/edu/wpi/first/shuffleboard/api/components/IntegerField.java]: should be isInt, here and throughout.

File: api/src/main/java/edu/wpi/first/shuffleboard/api/components/IntegerField.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+ */
+public class IntegerField extends TextField {
+
+  private final Property<Integer> number = new SimpleObjectProperty<>(this, "number", 0);
+  private static final Pattern numberPattern = Pattern.compile("^[-+]?\\d*$");
+
+  /**
+   * Creates a new number field with no value.
+   */
+  public IntegerField() {
```
Comment (nightpool) [inline api/src/main/java/edu/wpi/first/shuffleboard/api/components/IntegerField.java]: This is not the correct name or semantics for this class, since port numbers aren't integers, they're non-negative integers between the range of 0 and MAX_SHORT

File: app/src/main/java/edu/wpi/first/shuffleboard/app/prefs/AppPreferences.java [hunk 4]
```diff
@@ -34,7 +41,21 @@ public static AppPreferences getInstance() {
...
+        defaultTileSize,
+        server,
+        port
+    );
+  }
+
+  /**
+   * Gets a read-only list of all the preference properties.
+   */
+  public ImmutableList<Property<?>> getFlushableProperties() {
```
Comment (nightpool) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/prefs/AppPreferences.java]: this is a bad interface—it requires the caller to know whether it wants flushable properties or not. In reality, we want to mix and match the properties. For example, theme should change in real time, not just when you close the window.

File: app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java [hunk 5]
```diff
@@ -34,7 +37,32 @@ public void init() throws AlreadyLockedException, IOException {
...
+      } else {
+        NetworkTable.setIPAddress(newValue);
+      }
+    };
+    AppPreferences.getInstance().serverProperty().addListener(serverChangeListener);
+
+    ChangeListener<Integer> portChangeListener = (observable, oldValue, newValue) -> {
+      NetworkTable.shutdown();
+      if (newValue == null || newValue <= 0 || newValue >= 65535) {
+        NetworkTable.setPort(NetworkTable.DEFAULT_PORT);
```
Comment (nightpool) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java]: this is bad UX, since it doesn't update the UI when it sets it to the default value. this should be encapsulated into the field instead. (new PortField(), portField.setDefault(NetworkTable.DEFAULT_PORT) or something)

File: app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java [hunk 6]
```diff
@@ -34,7 +37,32 @@ public void init() throws AlreadyLockedException, IOException {
...
-    NetworkTablesJNI.startClient("localhost", 1735);
+    NetworkTable.setClientMode();
+
+    ChangeListener<String> serverChangeListener = (observable, oldValue, newValue) -> {
+      if (newValue.matches("[1-9](\\d{1,3})?")) {
+        NetworkTable.setTeam(Integer.parseInt(newValue));
+      } else if (newValue.isEmpty()) {
+        NetworkTable.setIPAddress("localhost");
+      } else {
+        NetworkTable.setIPAddress(newValue);
```
Comment (nightpool) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java]: I'm not sure i'm happy with the undiscoverable behavior here. I think ideally it would have a "team number" prompt and a hidden "hostname" field that you could show by clicking on a link, something like "Or, set the server directly...").

File: app/src/main/java/edu/wpi/first/shuffleboard/app/MainWindowController.java [hunk 7]
```diff
@@ -277,29 +276,29 @@ public void load() {
...
+    dialog.getDialogPane().getButtonTypes().addAll(ButtonType.CANCEL, ButtonType.OK);
+    dialog.setTitle("Shuffleboard Preferences");
+    dialog.setResizable(true);
+    dialog.setResultConverter(button -> !button.getButtonData().isCancelButton());
+    if (dialog.showAndWait().orElse(false)) {
+      propertySheet.getItems().stream()
+          .map(item -> (WidgetPropertySheet.PropertyItem) item)
+          .map(WidgetPropertySheet.PropertyItem::getObservableValue)
+          .filter(Optional::isPresent)
+          .map(property -> (FlushableProperty) property.get())
```
Comment (nightpool) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/MainWindowController.java]: unchecked cast. use map(o -> o.map(TypeUtils.optionalCast(FlushableProperty.class))) instead.

File: api/src/main/java/edu/wpi/first/shuffleboard/api/components/IntegerField.java (hunk 3 above)
Comment (AustinShalit) [inline api/src/main/java/edu/wpi/first/shuffleboard/api/components/IntegerField.java]: It is a Whole Number field.

File: api/src/main/java/edu/wpi/first/shuffleboard/api/components/IntegerField.java (hunk 2 above)
Comment (AustinShalit) [inline api/src/main/java/edu/wpi/first/shuffleboard/api/components/IntegerField.java]: Ok

File: app/src/main/java/edu/wpi/first/shuffleboard/app/prefs/AppPreferences.java (hunk 4 above)
Comment (AustinShalit) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/prefs/AppPreferences.java]: Ok

File: api/src/main/java/edu/wpi/first/shuffleboard/api/components/IntegerField.java (hunk 3 above)
Comment (SamCarlberg) [inline api/src/main/java/edu/wpi/first/shuffleboard/api/components/IntegerField.java]: `IntegerValueField` or `IntegralValueField` may be better names

File: app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java (hunk 6 above)
Comment (AustinShalit) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java]: Disagree.  A single field can support either a hostname or a team number.  The problem with having 2 fields is you can only use one at a time.

File: app/src/main/java/edu/wpi/first/shuffleboard/app/MainWindowController.java (hunk 7 above)
Comment (AustinShalit) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/MainWindowController.java]: ok

File: app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java (hunk 6 above)
Comment (nightpool) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java]: then the field should be labeled appropriately.

File: app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java (hunk 5 above)
Comment (AustinShalit) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java]: I am looking for guidance on how to include a PortField class into WidgetPropertySheet.  Right now we create fields based on their type.  A port field has the same type as an Integer field.

File: app/src/main/java/edu/wpi/first/shuffleboard/app/MainWindowController.java [hunk 8]
```diff
@@ -277,29 +276,32 @@ public void load() {
   @SuppressWarnings("unchecked")
   @FXML
   public void showPrefs() {
-    // Create the property sheet
-    PropertySheet propertySheet = new WidgetPropertySheet(AppPreferences.getInstance().getProperties());
+    PropertySheet propertySheet
+        = new WidgetPropertySheet(AppPreferences.getInstance().getProperties());
```
Comment (nightpool) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/MainWindowController.java]: unrelated formatting change?

File: app/src/main/java/edu/wpi/first/shuffleboard/app/MainWindowController.java [hunk 9]
```diff
@@ -277,29 +276,32 @@ public void load() {
...
+    dialog.getDialogPane().getButtonTypes().addAll(ButtonType.CANCEL, ButtonType.OK);
+    dialog.setTitle("Shuffleboard Preferences");
+    dialog.setResizable(true);
+    dialog.setResultConverter(button -> !button.getButtonData().isCancelButton());
+    if (dialog.showAndWait().orElse(false)) {
+      propertySheet.getItems().stream()
+          .map(item -> (WidgetPropertySheet.PropertyItem) item)
+          .map(WidgetPropertySheet.PropertyItem::getObservableValue)
+          .filter(Optional::isPresent)
+          .map(Optional::get)
```
Comment (nightpool) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/MainWindowController.java]: god I hate that java's Optional api doesn't implement the stream api... these three calls could just be one flatMap

File: app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java [hunk 10]
```diff
@@ -34,7 +38,34 @@ public void init() throws AlreadyLockedException, IOException {
...
+      /*
+       * You MUST set the port number before setting the team number.
+       */
+      if (value.length > 1) {
+        NetworkTable.setPort(Integer.parseInt(value[1]));
+      } else {
+        NetworkTable.setPort(NetworkTable.DEFAULT_PORT);
+      }
+
+      if (value[0].matches("[1-9](\\d{1,3})?")) {
```
Comment (nightpool) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java]: myhost:1191 matches, as does "my1991host"

File: app/src/main/java/edu/wpi/first/shuffleboard/app/MainWindowController.java (hunk 9 above)
Comment (SamCarlberg) [reply app/src/main/java/edu/wpi/first/shuffleboard/app/MainWindowController.java]: It's pretty easy to implement a version of Optional that implements `Stream`. Or we could use Kotlin and add an extension method 😉

File: app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java [hunk 11]
```diff
@@ -34,7 +38,34 @@ public void init() throws AlreadyLockedException, IOException {
...
+        NetworkTable.setTeam(Integer.parseInt(value[0]));
+      } else if (value[0].isEmpty()) {
+        NetworkTable.setIPAddress("localhost");
+      } else {
+        NetworkTable.setIPAddress(value[0]);
+      }
+
+      NetworkTable.initialize();
+    };
```
Comment (nightpool) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java]: needs a newline after

File: app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java [hunk 12]
```diff
@@ -34,7 +38,34 @@ public void init() throws AlreadyLockedException, IOException {
...
+    AppPreferences.getInstance().serverProperty().addListener(serverChangeListener);
+
```
Comment (nightpool) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/Shuffleboard.java]: move 65 down, add newline after changeListener

File: app/src/main/java/edu/wpi/first/shuffleboard/app/prefs/FlushableProperty.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+  private final Property<T> property;
+
+  /**
+   * Create a new FlushableProperty.
+   *
+   * @param property The property to flush to
+   */
+  public FlushableProperty(Property<T> property) {
+    super(null, property.getName(), property.getValue());
```
Comment (nightpool) [inline app/src/main/java/edu/wpi/first/shuffleboard/app/prefs/FlushableProperty.java]: `property.getBean();`

**Comment by SamCarlberg** [comment]
@nightpool you were too quick to approve this. Changing the team number doesn't clear NetworkTables, so every entry that was in the first server will be pushed to the new one.

This should also really be part of the NetworkTables plugin, but with alpha coming up this is OK for a minimum viable product
